R/response_process.R

```diff
@@ -0,0 +1,190 @@
+  if (httr::status_code(resp) == 204) {
+    return(TRUE)
+  }
+
+  check_response(resp)
+  content <- httr::content(resp, as = "raw")
+
+  if (raw) {
+    content
+  } else {
+    jsonlite::fromJSON(rawToChar(content), simplifyVector = FALSE)
+  }
+}
+
+check_response <- function(resp) {
+  code <- httr::status_code(resp)
+  # code could be in 100s, 200s, 300s, 400s, 500s
+
+  if (code < 200 || (code >= 300 && code < 400)) {
+    message <- c(
+      httr::http_status(resp)$message,
+      "  * HTTP status codes in the 100s and 300s are unexpected."
+    )
+    stop_request_failed(resp, message = message)
+  }
+  # code could be in 200s, 400s, 500s
+
+  if (code >= 200 && code < 300) {
+    check_for_json(resp)
+    return()
+  }
+  # code could be: 400-499, 500-599
+
+  google_error(resp)
+}
+
+stop_request_failed <- function(resp, ..., message = NULL, .subclass = NULL) {
+  code <- httr::status_code(resp)
+  class <- c(.subclass, glue("http_error_{code}"), "gargle_error_request_failed")
+
+  message <- message %||% httr::http_status(resp)$message
+  if (length(message) > 1) {
+    message <- glue_collapse(message, sep = "\n")
+  }
+
+  rlang::abort(
+    message,
+    .subclass = class,
+    code = code,
+    resp = redact_response(resp),
+    ...
+  )
+}
+
+check_for_json <- function(resp) {
+  type <- httr::http_type(resp)
+  if (!grepl("^application/json", type)) {
+    stop_need_json(resp)
+  }
+  invisible(resp)
+}
+
+stop_need_json <- function(resp) {
+  type <- httr::http_type(resp)
+  message <- c(
+    httr::http_status(resp)$message,
+    glue("  * Non-JSON content type: {type}")
+  )
+  stop_request_failed(
+    resp,
+    message = message,
+    text = httr::content(resp, as = "text")
+  )
+}
+
+google_error <- function(resp) {
+  check_for_json(resp)
+
+  code    <- httr::status_code(resp)
+  type    <- httr::http_type(resp)
+  content <- httr::content(resp, as = "raw")
+  content <- jsonlite::fromJSON(rawToChar(content), simplifyVector = FALSE)
+
+  error <- content[["error"]]
+  if(is.null(error)) {
+    # developed from test fixture from tokeninfo endpoint
+    message <- c(
+      httr::http_status(resp)$message,
+      glue("  * {content$error_description}")
+    )
+  } else {
+    errors <- error[["errors"]]
+    if (is.null(errors)) {
+      # developed from test fixture from "sheets.spreadsheets.get" endpoint
+      status <- httr::http_status(resp)
+      rpc <- rpc_description(error$status)
+      message <- c(
+        glue("{status$category}: ({error$code}) {error$status}"),
+        glue("  * {rpc}"),
+        glue("  * {error$message}")
+      )
+    } else {
+      # developed from test fixture from "drive.files.get" endpoint
+      errors <- unlist(errors)
+      message <- c(
+        httr::http_status(resp)$message,
+        glue("  * {format(names(errors), justify = 'right')}: {errors}")
+      )
+    }
+  }
+  stop_request_failed(resp, message = message)
+}
+
+redact_response <- function(resp) {
```

Comment:
I think the pros of this outweighs the cons. If people store a response as `.rds` (such as the test fixtures I'm adding), it seems best to redact the auth tokens.